### PR TITLE
Mapping for DP IDs

### PR DESCRIPTION
**Pull request by mschlenstedt**
Hi all,

I added a first "Proof of Concept" for doing a DPS <-> Name/Code mapping to have a more comfortable way to set and read the DPS.

At a first step I included mappings from IOBroker and FHEM project (they already use the DPS<->Code/Name mapping). The mapping is read once at startup after reading the devices.json and after saving a new deivces.json. Only those mappings are kept in memory (in a dict) which are currently used. This keeps things small and almost fast.

`/status/{DEVICE_ID}` now additionally give back the mapping scheme if one exists for the Product_Key in `mappings/mappings.json`:

```
{"devId": "05047xxxxxx923ca8", "dps": {"1": false, "2": 21, "3": 13, "4": "hot", "9": 256}, "dps_scheme": [{"id": 1, "code": "Power"}, {"id": 2, "code": "temp_set"}, {"id": 3, "code": "temp_current"}, {"id": 4, "code": "mode"}, {"id": 9, "code": "fault"}]}
```

At a first step I included the mapping in the WebUI:

image

What's still missing:
- Re-Read mappings if a device is discovered
- Add mappings in `/set/....` command

### CODE REVIEW

File: server/mappings/import.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# -*- coding: utf-8 -*-
+
+# Import DPS mappings from other projects and create mappings file
+
```
Comment (jasonacox) [inline server/mappings/import.py:4]: Suggestion: Add a comment block with author credit here! 😉 

```python
"""
TinyTuya Tool to Import DPS mappings from other projects and create mappings file

Author: @mschlenstedt 
Date: May 23, 2023
For more information see https://github.com/jasonacox/tinytuya
"""
```

**Comment by jasonacox** [comment]
Very cool @mschlenstedt !

@uzlonewolf - I know we don't want to  burden the library with a large json mapping payload (it is TINYtuya after all), but this seems reasonable for the server tool.  Thoughts? 😄

**Comment by mschlenstedt** [comment]
I agree for the Lib to not include mapping - this should be done on the application side. But as the server is an application, I would really like mapping here :-)

**Comment by mschlenstedt** [comment]
Grabbing the mapping "live" is cool - I wasn't successfull with it to be honest. But it's more "tiny" than delivering a 5 MB json database...

Important for me is that the /status call dilivers the scheme for other software who uses the API. I would like to use it from my MQTT script.

**Comment by mschlenstedt** [comment]
I played a little bit around with https://eu.iot.tuya.com/cloud/explorer But I haven't found any Cloud call which gives me back name/code and DP ID. Best results I found was with `https://openapi.tuyaeu.com/v1.2/iot-03/devices/{device_id}/specification` and/or `https://openapi.tuyaeu.com/v1.0/iot-03/devices/{device_id}/functions`. But bith just give me the names and codes, but not the corresponding DPS.

```
{
  "result": {
    "category": "rs",
    "functions": [
      {
        "code": "Power",
        "desc": "{}",
        "name": "开关",
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "temp_set",
        "desc": "{\"unit\":\"℃\",\"min\":8,\"max\":40,\"scale\":0,\"step\":1}",
        "name": "设置温度",
        "type": "Integer",
        "values": "{\"unit\":\"℃\",\"min\":8,\"max\":40,\"scale\":0,\"step\":1}"
      },
      {
        "code": "mode",
        "desc": "{\"range\":[\"auto\",\"cold\",\"hot\"]}",
        "name": "模式",
        "type": "Enum",
        "values": "{\"range\":[\"auto\",\"cold\",\"hot\"]}"
      }
    ],
    "status": [
      {
        "code": "Power",
        "name": "开关",
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "temp_set",
        "name": "设置温度",
        "type": "Integer",
        "values": "{\"unit\":\"℃\",\"min\":8,\"max\":40,\"scale\":0,\"step\":1}"
      },
      {
        "code": "temp_current",
        "name": "当前温度",
        "type": "Integer",
        "values": "{\"unit\":\"℃\",\"min\":-10,\"max\":100,\"scale\":0,\"step\":1}"
      },
      {
        "code": "mode",
        "name": "模式",
        "type": "Enum",
        "values": "{\"range\":[\"auto\",\"cold\",\"hot\"]}"
      },
      {
        "code": "fault",
        "name": "故障告警",
        "type": "Bitmap",
        "values": "{\"label\":[\"P1\",\"P2\",\"P3\",\"P4\",\"P6\",\"P7\",\"P8\",\"P9\",\"PL\",\"PC\",\"E3\",\"E4\",\"E8\",\"CS\",\"E6\"],\"maxlen\":15}"
      }
    ]
  },
  "success": true,
  "t": 1684949165736,
  "tid": "105ee88bfa5811edbfbfa24bf1d21c99"
}
```

**Comment by mschlenstedt** [comment]
Yes, I can confirm that I get same result as you with cloud.getdps(dev_id)`. This is great - never tried this!

Maybe you should use "code" instead of "name" in your mappings.json - as this is the standard. When your PR is included, I can change my functions to use the format if your mappings.json. 

```
{
    "result": {
        "category": "rs",
        "functions": [
            {
                "code": "Power",
                "dp_id": 1,
                "type": "Boolean",
                "values": "{}"
            },
            {
                "code": "temp_set",
                "dp_id": 2,
                "type": "Integer",
                "values": "{\"unit\":\"\u2103\",\"min\":8,\"max\":40,\"scale\":0,\"step\":1}"
            },
            {
                "code": "mode",
                "dp_id": 4,
                "type": "Enum",
                "values": "{\"range\":[\"auto\",\"cold\",\"hot\"]}"
            }
        ],
        "lang_config": {},
        "status": [
            {
                "code": "Power",
                "dp_id": 1,
                "type": "Boolean",
                "values": "{}"
            },
            {
                "code": "temp_set",
                "dp_id": 2,
                "type": "Integer",
                "values": "{\"unit\":\"\u2103\",\"min\":8,\"max\":40,\"scale\":0,\"step\":1}"
            },
            {
                "code": "temp_current",
                "dp_id": 3,
                "type": "Integer",
                "values": "{\"unit\":\"\u2103\",\"min\":-10,\"max\":100,\"scale\":0,\"step\":1}"
            },
            {
                "code": "mode",
                "dp_id": 4,
                "type": "Enum",
                "values": "{\"range\":[\"auto\",\"cold\",\"hot\"]}"
            },
            {
                "code": "fault",
                "dp_id": 9,
                "type": "Bitmap",
                "values": "{\"label\":[\"P1\",\"P2\",\"P3\",\"P4\",\"P6\",\"P7\",\"P8\",\"P9\",\"PL\",\"PC\",\"E3\",\"E4\",\"E8\",\"CS\",\"E6\"],\"maxlen\":15}"
            }
        ]
    },
    "success": true,
    "t": 1684958124149,
    "tid": "ec01691cfa6c11edbfbfa24bf1d21c99"
}
```

**Comment by mschlenstedt** [comment]
I now changed over to the DPS mapping from Tuya Cloud now. The server now uses the mappings found in devices.json (if any). It's shown in the WebUI and also used in the set API call, e. g.

```/set/pool%20heat%20pump/temp_set/26```

is possible now.

File: server/server.py [hunk 2]
```diff
@@ -102,6 +102,7 @@
 
 # Static Assets
 web_root = os.path.join(os.path.dirname(__file__), "web")
+MAPPINGFILE = os.path.join(os.path.dirname(__file__), "mappings", "mappings.json")
```
Comment (jasonacox) [inline server/server.py:105]: Should we remove this now?

File: server/server.py [hunk 3]
```diff
@@ -117,6 +118,7 @@
 running = True
 tuyadevices = []
 deviceslist = {}
+dpsmappings = {}
```
Comment (jasonacox) [inline server/server.py:121]: Remove?

**Comment by jasonacox** [comment]
This is great @mschlenstedt !!  Love this...

image

File: server/server.py (hunk 2 above)
Comment (jasonacox) [reply server/server.py]: Removed

File: server/server.py (hunk 3 above)
Comment (jasonacox) [reply server/server.py]: Removed

**Comment by jasonacox** [comment]
Thanks @mschlenstedt !!  This is a great enhancement to the server!

**Comment by mschlenstedt** [comment]
Thanks for removing the obsolete stuff ;-)

**Pull request merged**